clients/roscpp/include/ros/timer_manager.h

```diff
@@ -428,11 +428,36 @@ void TimerManager<T, D, E>::setPeriod(int32_t handle, const D& period)
   {
     return;
   }
+  
+  if(reset)
+  {
+    info->next_expected = T::now() + period;
+  }
+  
+  // If called in a callback, last_real has not been updated yet so (now - last_real) > period
+  // In this case, let next_expected be updated only in updateNext
```

Comment:
The issue with this is that if the callback calls setPeriod on its own timer, ```info->last_real``` is not updated by that point - it still holds the value of the previous callback. That seems like unintended behavior though.

Using the logic you posted, the following happens:
If ```setPeriod``` is called on the very first callback, it works fine:

>t=0 create Timer of 2s
>t=2 callback occurs, call setPeriod(3)
>t=5 callback occurs
>t=8 callback occurs

If ```setPeriod``` is called on the second callback:
>t=0 create Timer of 2s
>t=2 callback occurs
>t=4 callback occurs, call setPeriod(3)
>t=5 callback occurs
>t=8 callback occurs